### PR TITLE
594 bug fix baseline johntheripper

### PR DESCRIPTION
**Pull request by slpysiren**
Completed the baseline checklist for the component. Please let me know if I missed anything or need to make any changes.

### CODE REVIEW

File: src/components/JohnTheRipper/JohnTheRipper.tsx [hunk 1]
```diff
@@ -1,117 +1,188 @@
...
+    const [loadingModal, setLoadingModal] = useState(true); // State variable to indicate loading state of the modal.
+    const [isCommandAvailable, setIsCommandAvailable] = useState(false); // State variable to check if the command is available.
+    const [opened, setOpened] = useState(!isCommandAvailable); // State variable that indicates if the modal is opened.
+    const [allowSave, setAllowSave] = useState(false); // State variable to allow saving the output to a file.
+    const [hasSaved, setHasSaved] = useState(false); // State variable to indicate if the output has been saved.
+
+    // Component constants
+    const [selectedFileTypeOption, setSelectedFileTypeOption] = useState(""); // State variable to store the selected file type.
+    const [selectedModeOption, setSelectedModeOption] = useState(""); // State variable to store the selected crack mode.
+    const [selectedIncrementOption, setSelectedIncrementOption] = useState(""); // State variable to store the selected increment order.
```
Comment (bradleybogg) [inline src/components/JohnTheRipper/JohnTheRipper.tsx:42]: These are all react state hooks as indicated by the "useState". Please move them into that appropriate section. 
```suggestion
```

File: src/components/JohnTheRipper/JohnTheRipper.tsx [hunk 2]
```diff
@@ -1,117 +1,188 @@
...
-    let form = useForm({
+    // Component state variables
+    const [loading, setLoading] = useState(false); // State variable to indicate loading state.
+    const [output, setOutput] = useState(""); // State variable to store the output of the command execution.
+    const [pid, setPid] = useState(""); // State variable to store the process ID of the command execution.
+    const [loadingModal, setLoadingModal] = useState(true); // State variable to indicate loading state of the modal.
+    const [isCommandAvailable, setIsCommandAvailable] = useState(false); // State variable to check if the command is available.
+    const [opened, setOpened] = useState(!isCommandAvailable); // State variable that indicates if the modal is opened.
+    const [allowSave, setAllowSave] = useState(false); // State variable to allow saving the output to a file.
+    const [hasSaved, setHasSaved] = useState(false); // State variable to indicate if the output has been saved.
```
Comment (bradleybogg) [inline src/components/JohnTheRipper/JohnTheRipper.tsx:37]:
```suggestion
    const [hasSaved, setHasSaved] = useState(false); // State variable to indicate if the output has been saved.
    const [selectedFileTypeOption, setSelectedFileTypeOption] = useState(""); // State variable to store the selected file type.
    const [selectedModeOption, setSelectedModeOption] = useState(""); // State variable to store the selected crack mode.
    const [selectedIncrementOption, setSelectedIncrementOption] = useState(""); // State variable to store the selected increment order.

```

File: src/components/JohnTheRipper/JohnTheRipper.tsx [hunk 3]
```diff
@@ -1,117 +1,188 @@
...
+    // Component constants
+    const [selectedFileTypeOption, setSelectedFileTypeOption] = useState(""); // State variable to store the selected file type.
+    const [selectedModeOption, setSelectedModeOption] = useState(""); // State variable to store the selected crack mode.
+    const [selectedIncrementOption, setSelectedIncrementOption] = useState(""); // State variable to store the selected increment order.
+    const modeRequiringWordList = ["dictionary"]; // Crack modes that require a wordlist
+    const modeRequiringIncrementOrder = ["incremental"]; // Crack modes that require an increment order
+    const fileTypes = ["zip", "rar", "raw"]; // File types supported by the tool
+    const mode = ["incremental", "dictionary", "single"]; // Crack modes supported by the tool
+    const incrementOrder = [
+        // Increment orders
```
Comment (bradleybogg) [inline src/components/JohnTheRipper/JohnTheRipper.tsx:48]: Remove this comment.

File: src/components/JohnTheRipper/JohnTheRipper.tsx [hunk 4]
```diff
@@ -1,117 +1,188 @@
...
+        "Upper",
+        "Digits",
+        "LM_ASCII",
+    ];
+    const title = "John the Ripper"; // Title of the component.
+    const description =
+        "John the Ripper is a fast password cracker, its primary purpose is to detect weak Unix passwords."; // Description of the component.
+    const steps =
+        "Step 1: Specify the filepath to the password file that you wish to crack (e.g ~/passwords.txt).\n" +
+        "Step 2: Specify the hashing algorithm used by the password you are trying to crack (e.g md5). For a full list of options, open new terminal and use john --list=formats command.\n" +
```
Comment (bradleybogg) [inline src/components/JohnTheRipper/JohnTheRipper.tsx:66]:
```suggestion
        "Step 2: Specify the hashing algorithm used by the password you are trying to crack (e.g md5).\n" +
```

File: src/components/JohnTheRipper/JohnTheRipper.tsx [hunk 5]
```diff
@@ -1,117 +1,188 @@
...
+        "Step 3: Specify which cracking mode to use.\n" +
```
Comment (bradleybogg) [inline src/components/JohnTheRipper/JohnTheRipper.tsx:67]:
```suggestion
        "Step 3: Specify the cracking mode to use.\n" +
```

File: src/components/JohnTheRipper/JohnTheRipper.tsx [hunk 6]
```diff
@@ -1,117 +1,188 @@
...
+        "Step 4: Select the file type of the password file.\n" +
```
Comment (bradleybogg) [inline src/components/JohnTheRipper/JohnTheRipper.tsx:68]:
```suggestion
        "Step 4: Select the file type.\n" +
```

File: src/components/JohnTheRipper/JohnTheRipper.tsx [hunk 7]
```diff
@@ -1,117 +1,188 @@
...
+        "Step 5: If you selected the Incremental or Dictionary cracking type, an additional option will appear to select a character set or specify a filepath for a wordlist respectively." +
+        "Step 6: Click crack to commence the tool's execution.\n" +
```
Comment (bradleybogg) [inline src/components/JohnTheRipper/JohnTheRipper.tsx:70]:
```suggestion
        "Step 6: Click 'Start John the Ripper'.\n" +
```

File: src/components/JohnTheRipper/JohnTheRipper.tsx [hunk 8]
```diff
@@ -1,117 +1,188 @@
...
+        "Step 7: View the output block to view the results of the tool's execution.\n";
```
Comment (bradleybogg) [inline src/components/JohnTheRipper/JohnTheRipper.tsx:71]:
```suggestion
        "Step 7: View the output block to view the results of the tools execution.\n";
```

File: src/components/JohnTheRipper/JohnTheRipper.tsx [hunk 9]
```diff
@@ -137,9 +209,9 @@ const JohnTheRipper = () => {
             //Crack the extracted hash
```
Comment (bradleybogg) [inline src/components/JohnTheRipper/JohnTheRipper.tsx:209]:
```suggestion
            // Crack the extracted hash
```

File: src/components/JohnTheRipper/JohnTheRipper.tsx [hunk 10]
```diff
@@ -137,9 +209,9 @@ const JohnTheRipper = () => {
...
             values.hash ? argsCrack.push(`--format=${values.hash}`) : undefined;
             selectedModeOption === "dictionary"
-                ? argsCrack.push(`--wordlist=${values.wordlist}`)
+                ? argsCrack.push(`--wordlist=${values.wordList}`)
                 : selectedModeOption === "incremental"
-                ? argsCrack.push(`-incremental:${values.incrementorder}`)
+                ? argsCrack.push(`-incremental:${values.incrementOrder}`)
                 : argsCrack.push(`--single`);
 
             try {
```
Comment (bradleybogg) [inline src/components/JohnTheRipper/JohnTheRipper.tsx:217]: Update this block to use the overlay for cancelling

**Review by bradleybogg — CHANGES_REQUESTED** [review]
Very good work @slpysiren

I appreciate your efforts in doing it right and taking the time to do so.
Good work on this one. I have some very minor changes to make. Some of the comments for steps I've added reccomendations. See which you think is better. For the others please implement changes.

Thank you!

**Review by michaeljpigott — APPROVED** [review]
Hi @slpysiren 

This baselining work looks good. I approve to merge pending app testing from @zkkok 

Kind regards

Michael

File: src/components/JohnTheRipper/JohnTheRipper.tsx [hunk 11]
```diff
@@ -1,131 +1,213 @@
 import { Button, LoadingOverlay, NativeSelect, Stack, TextInput } from "@mantine/core";
```
Comment (zkkok) [inline src/components/JohnTheRipper/JohnTheRipper.tsx:1]: Remove unused imports

File: src/components/JohnTheRipper/JohnTheRipper.tsx [hunk 12]
```diff
@@ -1,131 +1,213 @@
...
-    "\n\nStep 6: Click crack to commence the tool's execution." +
-    "\n\nStep 7: View the output block below to view the results of the tool's execution.";
+import { SaveOutputToTextFile_v2 } from "../SaveOutputToFile/SaveOutputToTextFile";
+import { RenderComponent } from "../UserGuide/UserGuide";
+import InstallationModal from "../InstallationModal/InstallationModal";
+import { checkAllCommandsAvailability } from "../../utils/CommandAvailability";
+import { LoadingOverlayAndCancelButton } from "../OverlayAndCancelButton/OverlayAndCancelButton";
 
+/**
+ * Represents the form values for the AirbaseNG component.
```
Comment (zkkok) [inline src/components/JohnTheRipper/JohnTheRipper.tsx:14]: Revise comment

**Review by zkkok — CHANGES_REQUESTED** [review]
Hi @slpysiren, Application testing has been done. There's a few issues but most of them pre-existed within the component. Only need to make some minor changes before it can be merged.

Summary of functionality test:
[Functionlity Testing(John).pdf](https://github.com/Hardhat-Enterprises/Deakin-Detonator-Toolkit/files/15282610/Functionlity.Testing.John.pdf)

**Comment by michaeljpigott** [comment]
Hi @zkkok, I will let you review your requested changes for this one. 

Thanks.

**Review by zkkok — APPROVED** [review]
Baseline bug fix looks good without any issue in sight.

**Pull request merged**